src/R/Core/Impl/AST/Extensions/ScopeExtensions.cs

```diff
@@ -0,0 +1,129 @@
+
+namespace Microsoft.R.Core.AST {
+    public static class ScopeExtensions {
+        /// <summary>
+        /// Enumerates scopes from the current up to the global scope.
+        /// </summary>
+        public static IEnumerable<IScope> EnumerateTowardsGlobal(this IScope scope) {
+            yield return scope;
+            while (!(scope is GlobalScope)) {
+                scope = scope.GetEnclosingScope();
+                yield return scope;
+            }
+        }
+
+        /// <summary>
+        /// Enumerates definitions of variables applicable to the given scope.
+        /// Traverses scopes from the provided scope up and enumerates
+        /// variables that appear at the left side of the left-hand assignment
+        /// operator or at the right side of the right-hand assignmend operator.
+        /// Includes both regular variables as well as function definitions.
+        /// Includes assignments that appear in the file up to the specified
+        /// position except in the global scope it enumerates all assignments.
+        /// </summary>
+        /// <param name="position"></param>
+        public static IEnumerable<IVariable> GetApplicableVariables(this IScope scope, int position) {
+            foreach (var s in scope.EnumerateTowardsGlobal()) {
+                foreach (var v in s.GetScopeVariables(position)) {
+                    yield return v;
+                    position = s.Start;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Enumerates definitions of variables applicable to the given scope.
+        /// Enumerates all the variables that appear at the left side of 
+        /// the left assignment operator or at the right side of the right-hand 
+        /// assignmend operator. Includes both regular variables as well as 
+        /// function definitions.
+        /// </summary>
+        /// <param name="scope">Scope to look into</param>
+        /// <param name="position">
+        /// If scope is not a global scope, then only variables before this position
+        /// will be enumerated. In global scope all variables are enumerated. This
+        /// reflects how variables are visible when file is sources and user types
+        /// somewhere in inner scope.
+        /// </param>
+        public static IEnumerable<IVariable> GetScopeVariables(this IScope scope, int position) {
+            bool globalScope = scope is GlobalScope;
+
+            if (!globalScope) {
+                // See if this is a function scope with arguments
+                var funcDef = scope.Parent as IFunctionDefinition;
+                if (funcDef != null && funcDef.Arguments != null) {
+                    foreach (var arg in funcDef.Arguments) {
+                        var na = arg as IVariable;
+                        if (na != null) {
+                            yield return na;
+                        } else {
+                            var ea = arg as ExpressionArgument;
```

Comment:
ExpressionArgument is a most common type of arguments. i.e. in func(x+1, a) both are expression arguments. 'x = 1' is a named argument. ,, is a missing argument, in func(,, third argument is a 'stub argument'.